src/index/blockfilter.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <map>
+
+#include <dbwrapper.h>
+#include <index/blockfilter.h>
+#include <util.h>
+#include <validation.h>
+
+/* The index database stores three items for each block: the encoded filter, its dSHA256 hash, and
+ * the header. Those belonging to blocks on the active chain are indexed by height, and those
+ * belonging to blocks that have been reorganized out of the active chain are indexed by block hash.
+ * This ensures that filter data for any block that becomes part of the active chain can always be
+ * retrieved, alleviating timing concerns.
```

Comment:
Can you explain this? What would be wrong with always indexing by block hash? Especially given that `getblockfilter` takes a block hash argument.

Reply:
Indexing by hash is much slower when fetching a range of filters or filter hashes, which is a common access pattern in BIP 157.